### PR TITLE
Adds support for retrieving credentials in the Cloud Shell environment.

### PR DESCRIPTION
**Pull request by wilford**

### CODE REVIEW

**Comment by ejona86** [comment]
@anthmgoogle, are you going to review this one?

**Comment by anthmgoogle** [comment]
Yes, I'll review this today.

File: oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * OAuth2 credentials representing the built-in service account for Google Cloud Shell.
+ *
+ * <p>Fetches access tokens from the App Identity service.
+ */
+public class CloudShellCredentials extends GoogleCredentials {
+
+  private final static int ACCESS_TOKEN_INDEX = 2;
+  private final static int READ_TIMEOUT = 5000;
```
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java]: Perhaps add the units for readability here?

File: oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  protected final static String GET_AUTH_TOKEN_REQUEST = "2\n[]";
```
Comment (anthmgoogle) [inline oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java]: There might be opportunity to improve this name to clarify what the content means.

**Comment by anthmgoogle** [comment]
Sorry for the delay. LGTM. A couple of minor naming suggestions.

**Comment by anthmgoogle** [comment]
One more thought: it may be worth porting this to google-api-client-java as well since that library currently  also has an ADC implementation, but will have broader reach for a while.

**Comment by aalexand** [comment]
> One more thought: it may be worth porting this to google-api-client-java as well since that library currently also has an ADC implementation, but will have broader reach for a while.

It looks like Dataflow SDK (which is the one we want to make sure working in the Cloud Shell now) uses com.google.api.client.googleapis.auth.oauth2.GoogleCredential, see https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/main/java/com/google/cloud/dataflow/sdk/util/Credentials.java#L24.  That one is from google-api-client-java I believe?

Can one implementation just call into the other so that no code duplication is needed?

**Comment by anthmgoogle** [comment]
Yes, most existing usage is google-api-client-java. Eventually the auto
code in google-auth-library-java will replace it.

Can one implementation just call into the other so that no code duplication
is needed?


This varies depending on the language. For Java, while it would be
theoretically possible to write a clunky wrapper to convert on credential
type into another, given that there is already an ADC implementation, I
would recommend just port it to both places. They both have a similar
structure in the implementation, so it should be relatively low cost to
port.

File: oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java (hunk 1 above)
Comment (wilford) [inline oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java]: Done

File: oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java (hunk 2 above)
Comment (wilford) [inline oauth2_http/java/com/google/auth/oauth2/CloudShellCredentials.java]: Added a comment that explains the formatting.

**Comment by wilford** [comment]
Added comments and updated name. 

I'll update google-api-client-java separately.

**Comment by anthmgoogle** [comment]
LGTM.

**Comment by anthmgoogle** [comment]
Let me know when you are ready for me to merge in the pull request.

**Comment by wilford** [comment]
We're good to merge.  Thanks!

**Pull request merged**